overlord/devicestate/devicestate.go

```diff
@@ -637,3 +640,61 @@ func Remodeling(st *state.State) bool {
 	}
 	return false
 }
+
+type recoverySystemSetup struct {
+	// Label of the recovery system, selected when tasks are created
+	Label string `json:"label"`
+	// Directory inside the seed filesystem where the recovery system files
+	// are kept, typically /run/mnt/ubuntu-seed/systems/<label>, set when
+	// tasks are created
+	Directory string `json:"directory"`
+	// SnapSetupTasks is a list of tasks that carry snap setup information,
+	// relevant only during remodel, set when tasks are created
+	SnapSetupTasks []string `json:"snap-setup-tasks"`
+}
+
+func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []string) (*state.TaskSet, error) {
+	// sanity check, the directory should not exist yet
+	// TODO: we should have a common helper to derive this path
+	systemDirectory := filepath.Join(boot.InitramfsUbuntuSeedDir, "systems", label)
+	exists, _, err := osutil.DirExists(systemDirectory)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return nil, fmt.Errorf("recovery system %q already exists", label)
+	}
+
+	create := st.NewTask("create-recovery-system", fmt.Sprintf("Create recovery system with label %q", label))
+	// the label we want
+	create.Set("recovery-system-setup", &recoverySystemSetup{
+		Label:     label,
+		Directory: systemDirectory,
+		// IDs of the tasks carrying snap-setup
+		SnapSetupTasks: snapSetupTasks,
+	})
+
+	finalize := st.NewTask("finalize-recovery-system", fmt.Sprintf("Finalize recovery system with label %q", label))
+	finalize.WaitFor(create)
+	// finalize needs to know the label too
+	finalize.Set("recovery-system-setup-task", create.ID())
+	return state.NewTaskSet(create, finalize), nil
+}
+
+func CreateRecoverySystem(st *state.State, label string) (*state.Change, error) {
+	var seeded bool
+	err := st.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	if !seeded {
+		return nil, fmt.Errorf("cannot remodel until fully seeded")
```

Comment:
s/remodel/create new recovery systems/ ?

Reply:
this check is not tested fwiw

Reply:
test added